### ISSUE TITLE
Refine nullability checks on interface implementations

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#38560]
Implementing the same interface twice, indirectly and with nullability differences involving oblivious currently produces a warning, but arguably shouldn't:
```C#
#nullable enable

public class C : I<object>, IObjectOblivious // warning CS8645: 'I<object>' is already listed in the interface list on type 'C' with different nullability of reference types.
{
}

public interface I<T> { }

#nullable disable

public interface IObjectOblivious : I<object> { }
```
([sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4FgBQBiAdgVwDZ4TB5wAEcWRJ66AAgMym0BMpAwqSKQJIA8A9sABWcAMYwAfABoeAeWFiY8vAEsAbiv44opAPS7SAdwgIsKrAHN2AZQAcANgAsAVi4ByPoJHiJb0ip0IPAQ4CAATAE9SVVg4MP8sUhgACzJzeAQAMwhRMhiYUn5EmAiABzI3Nj9DFRTSMJVMzMQKAtwCIhVVEsLM0hDmkKxcpLK4KAA6dABvdABfGjQGBIzskb4AFQlSadIFtHRsfEJiMgaoKjhF5fTENbJueW8lYnVNbU4eAQUfHb2gA==))

On the other hand, implementing the same interface twice, directly and with nullability differences arguably should produce an error instead of just a warning:
```C#
#nullable enable
public class C : I<object>, I<object?> // warning CS8645: 'I<object?>' is already listed in the interface list on type 'C' with different nullability of reference types.
{
}

public interface I<T> { }
```
([sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFAACAzIdQEyEDChIhAkgDwD2wAKzgBjGAD4ANFz6CRMAPxjCAemWEA7hARoAlmgDmrAMoAOAGwAWAKwcA5D35DRi24R1RCELAjgQAJgCehFju8H5uaIQwABZEevAIAGYQwkQhsIS8kTABAA5Etiyu6joxhH46iYmIJDCEmDh4OiE5mYmEPtU+aKlReXBQAHRUAN5UAL5UVHQRCcm9PAAqSiOE40A))

Finally, this behavior is not consistently enforced on type constraints: direct scenarios (`where T : I<object?>, I<object>`) produce an error and indirect scenarios (`where T : I<object?>, IObjectUnannotated`) produce no diagnostic.

Relates to tuple name checks on interface implementations (issue https://github.com/dotnet/roslyn/issues/38427 tracks some irregularities).
Relates to PR https://github.com/dotnet/roslyn/pull/38460 (type inference can now handle types that implement duplicate interfaces with nullability differences).
